### ISSUE TITLE
how to get all task status?

### ISSUE THREAD

**Issue by timzaak** [BinChengZhao/delay-timer#4]
If I should listen the status reporter message, and manager the status by myself?

**Comment by BinChengZhao** [comment]
All states of the task are maintained by delay-timer's several internal components and do not need to be maintained manually by the user.

The reason `status-reporter`  provides this is for two reasons: the user senses the passage of internal events and the need to interrupt a running task in advance ( interrupt as opposed to removal).

The PublicEvent provided by `StatusReporter` or `DelayTImer` is just the right handle to be able to cancel a running instance.（A task may be running and generate many running instances）

The current API is not flexible enough, and I noticed this, so v0.3.0 or v0.4.0 , the api for `add_tasks` will return a `Future` handle, which supports asynchronous fetching of internal events and asynchronous cancellation of running instances of tasks.

I hope my reply will answer your question.

**Comment by BinChengZhao** [comment]
> If I should listen the status reporter message, and manager the status by myself?

All states of the task are maintained by delay-timer's several internal components and do not need to be maintained manually by the user.

The reason status-reporter provides this is for two reasons: the user senses the passage of internal events and the need to interrupt a running task in advance ( interrupt as opposed to removal).

The PublicEvent provided by StatusReporter or DelayTImer is just the right handle to be able to cancel a running instance.（A task may be running and generate many running instances）

The current API is not flexible enough, and I noticed this, so v0.3.0 or v0.4.0 , the api for add_tasks will return a Future handle, which supports asynchronous fetching of internal events and asynchronous cancellation of running instances of tasks.

I hope my reply will answer your question.

**Comment by timzaak** [comment]
thanks.

**Comment by timzaak** [comment]
`CandyCronStr(&'static str s)`  should be `CandyCronStr(String s)`. otherwise can not construct CandyCronStr from param. For example:
```rust 
struct A{cron_str:String}

fun construct_candy(a:A) -> CandyCronStr{
   CandyCronStr(&a.cron_str.clone()) // this is not correct. and may use Box::leak to construct CandyCronStr
}
```

**Comment by BinChengZhao** [comment]
> `CandyCronStr(&'static str s)` should be `CandyCronStr(String s)`. otherwise can not construct CandyCronStr from param. For example:
> 
> ```rust
> struct A{cron_str:String}
> 
> fun construct_candy(a:A) -> CandyCronStr{
>    CandyCronStr(&a.cron_str.clone()) // this is not correct. and may use Box::leak to construct CandyCronStr
> }
> ```

You are right, the data given by the user is uncertain and I can't force a constraint requiring a 'static str, which is not flexible enough.

Just like if the user wanted to convert from UserCron(&'a str) or UserCron(String), the current type would cause problems for the user.

Thanks for the heads up, this issue recently followed the stable-to-`stable-rustc` compilation, together with the release (about 5-7 business days).

**Comment by timzaak** [comment]
fixed in feature/stable branch

**Comment by BinChengZhao** [comment]
> fixed in feature/stable branch

I have just released v0.3.0 with the following updates:

- Stable to stable-rustc compilation, with repair optimization.

1.Compilable at stable by conditional compilation.

2.Balancing performance and user experience (ajust CandyCronStr inner-type and add free-api for TaskBuilder).

3.Support custom setting of time zone.

4.Fix the clock too fast issue.

5.Use `next_second_hand` to solve a schedule problem.





And some plan for next version :

  // Here is an expected implementation for DelayTimer::add_task , which is not yet determined.
  **Do you have a better opinion?**

  ///```
  // let delay_timer = DelayTimer::default();
  //
  // let join_handle = delay_timer.add_task(_).unwrap();
  //
  // let peek : Option<Peek<&Instance>> = join_handle.peek().await;
  // let peek : Result<Option<Peek<&Instance>>> join_handle.try_peek();
  //
  // let instance : Option<Intance> =  join_handle.next().await.unwrap();
  // let instance : Result<Option<Intance>> =  join_handle.try_next().unwrap();
  //
  // instance.cancel();
  //
  ///```

**Comment by BinChengZhao** [comment]
> > fixed in feature/stable branch
> 
> I have just released v0.3.0 with the following updates:
> 
> * Stable to stable-rustc compilation, with repair optimization.
> 
> 1.Compilable at stable by conditional compilation.
> 
> 2.Balancing performance and user experience (ajust CandyCronStr inner-type and add free-api for TaskBuilder).
> 
> 3.Support custom setting of time zone.
> 
> 4.Fix the clock too fast issue.
> 
> 5.Use `next_second_hand` to solve a schedule problem.
> 
> And some plan for next version :
> 
> // Here is an expected implementation for DelayTimer::add_task , which is not yet determined.
> **Do you have a better opinion?**
> 
> ///` // let delay_timer = DelayTimer::default(); // // let join_handle = delay_timer.add_task(_).unwrap(); // // let peek : Option<Peek<&Instance>> = join_handle.peek().await; // let peek : Result<Option<Peek<&Instance>>> join_handle.try_peek(); // // let instance : Option<Intance> = join_handle.next().await.unwrap(); // let instance : Result<Option<Intance>> = join_handle.try_next().unwrap(); // // instance.cancel(); // ///`



> fixed in feature/stable branch

  // Here is an expected implementation for DelayTimer::add_task , which is not yet determined.
  **Do you have a better opinion?**

  ///```
  // let delay_timer = DelayTimer::default();
  //
  // let join_handle = delay_timer.add_task(_).unwrap();
  //
  // let peek : Option<Peek<&Instance>> = join_handle.peek().await;
  // let peek : Result<Option<Peek<&Instance>>> join_handle.try_peek();
  //
  // let instance : Option<Intance> =  join_handle.next().await.unwrap();
  // let instance : Result<Option<Intance>> =  join_handle.try_next().unwrap();
  //
  // instance.cancel();
  //
  ///```

**Comment by BinChengZhao** [comment]
> fixed in feature/stable branch

    // Here is an expected implementation, which is not yet determined.

    ///```

    // let delay_timer = DelayTimer::default();

    // let join_handle = delay_timer.add_task(_).unwrap();

    // let peek : Option<Peek<&Instance>> = join_handle.peek().await;

    // let peek : Result<Option<Peek<&Instance>>> join_handle.try_peek();

    // let instance : Option<Intance> =  join_handle.next().await.unwrap();

    // let instance : Result<Option<Intance>> =  join_handle.try_next().unwrap();

    // instance.cancel();

    ///```

**Comment by BinChengZhao** [comment]
Hi, @timzaak 

The just released v0.4.0 version already supports getting task instance & status from task handles.
